config/checkstyle-checks.xml

```diff
@@ -115,6 +115,13 @@
     <property name="matchAcrossLines" value="true" />
     <property name="message" value="Empty file is not allowed" />
   </module>
+  <module name="RegexpMultiline">
+    <property name="id" value="regexpCheckMessageKeyAnnotation"/>
+    <property name="format"
+              value="(?&lt;!@MessageKey)\n\s*public\sstatic\sfinal\sString\s.*MSG.*\s=\s&quot;.+&quot;;\n"/>
```

Comment:
Regex doesn't work if split in 2 lines.

Reply:
Seems like we could use MatchXPath check to make this easier/more reliable

Reply:
Oh I saw your other comment.

Reply:
better to try xpath, it might not be easier, but lets try  and choose.

Reply:
MatchXpath:
```
➜  src cat Test.java
public class Test {
    @MessageKey
    public static final String EXCEPTION_MSG = "general.exception";

    public static final String OTHER_EXCEPTION_MSG = "other.general.exception";

    @interface MessageKey { }
}%                                                                                                         ➜  src cat Test.java
public class Test {
    @MessageKey
    public static final String EXCEPTION_MSG = "general.exception";

    public static final String OTHER_EXCEPTION_MSG = "other.general.exception";

    @interface MessageKey { }
}

➜  src cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
    "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
    "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <property name="severity" value="error"/>
  <property name="fileExtensions" value="java, properties, xml"/>
  <module name="TreeWalker">
    <module name="MatchXpath">
      <property name="query" value="/COMPILATION_UNIT/CLASS_DEF/OBJBLOCK/VARIABLE_DEF[
        ./MODIFIERS/LITERAL_PUBLIC
        and ./MODIFIERS/LITERAL_STATIC
        and ./MODIFIERS/FINAL
        and not (./MODIFIERS/ANNOTATION/IDENT[@text='MessageKey'])]"/>
      <message key="matchxpath.match"
              value="Message keys should have @MessageKey annotation"/>
    </module>
 </module>
</module>
                                                                                                 ➜  src java -jar checkstyle-10.7.0-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /home/nick/IdeaProjects/tester/src/Test.java:5:5: Message keys should have @MessageKey annotation [MatchXpath]
Audit done.
Checkstyle ends with 1 errors.
```

Reply:
This should do it:
```
    <module name="MatchXpath">
      <property name="query" value="/COMPILATION_UNIT/CLASS_DEF/OBJBLOCK/VARIABLE_DEF[
        ./MODIFIERS/LITERAL_PUBLIC
        and ./MODIFIERS/LITERAL_STATIC
        and ./MODIFIERS/FINAL
        and ./IDENT[ends-with(@text,'MSG')]
        and not (./MODIFIERS/ANNOTATION/IDENT[@text='MessageKey'])]"/>
      <message key="matchxpath.match"
              value="Message keys should have @MessageKey annotation"/>
    </module>
```

Reply:
Done. Changed `ends-with` to `contains`. Most message key variables names start with `MSG`.